File: spark/src/main/scala/me/kosik/interwalled/ailist/Interval.scala
package me.kosik.interwalled.ailist

case class BucketedInterval(
  bucket:   String,
  key:      String,
  from:     Long,
  to:       Long,
  value:    String
) {

  def toInterval: Interval =
    Interval(key, from, to, value)
}

object IntervalColumns {
  val BUCKET : String = "bucket"
  val KEY    : String = "key"
  val FROM   : String = "from"
  val TO     : String = "to"
  val VALUE  : String = "value"

  // Additional, temporary columns
  val _COMPONENT  : String = "_ailist_component"
  val _MAX_E      : String = "_ailist_max_end"
}

File: ailist-core/src/main/java/me/kosik/interwalled/ailist/core/model/Interval.java
package me.kosik.interwalled.ailist.core.model;

public record Interval(
    long from,
    long to
) {

    public static boolean overlaps(final Interval lhs, final Interval rhs) {
        return lhs.from <= rhs.to && rhs.from <= lhs.to;
    }
}
